contrib/python/pypodman/pypodman/lib/parser_actions.py

```diff
@@ -40,22 +43,44 @@ def __init__(self,
                  choices=None,
                  required=False,
                  help=None,
-                 metavar='{True,False}'):
+                 metavar='True,False'):
         """Create BooleanAction object."""
+        if const is None or default is None:
+            raise ValueError(
+                "BooleanAction requires both const and default arguments.")
+
+        if help:
+            if 'default' not in help:
+                help += ' (default: %(default)s)'
+        else:
+            help = 'Default: %(default)s.'
+
+        # NOTE: values overrode here to ensure __call__() magic works.
         super().__init__(
             option_strings=option_strings,
             dest=dest,
-            nargs=nargs,
+            nargs=argparse.OPTIONAL,
             const=const,
             default=default,
-            type=type,
+            type=None,
             choices=choices,
             required=required,
             help=help,
             metavar=metavar)
 
     def __call__(self, parser, namespace, values, option_string=None):
-        """Convert and Validate input."""
+        """Convert and Validate input.
+
+        If option_string not given on CLI:
+            values = self.default
+        elsif option_string given but no value:
+            values = self.const
+        else:
+            values = value given on CLI
+        """
+        if values is None:
+            values = self.const
+
         try:
             val = BooleanValidate()(values)
         except ValueError:
```

Comment:
I think the crux of the matter is here: perhaps `ValueError` needs to be treated the same as 'no value given, use default'. As it is right now, I still need to give values when the option isn't the last one on the line:
```
# pypodman run --rm true centos:7 date
(hangs, as reported earlier)

# pypodman run --rm centos:7 date
2018-11-28 22:57:49 UTC | ERROR    | pypodman run: "--rm" must be True or False.
2018-11-28 22:57:49 UTC | ERROR    | Try 'pypodman run --help' for more information.
```

Unfortunately this requires pushing `values` back on the argparse arg list, and I can't figure out how to do that.

Reply:
My testing used ```pypodman commit --quiet --change Label=blue=red awesome_hypatia  test_commit```

The call to --change "reset" the parser. :(  ```pypodman run --rm -- centos:7 date``` parses as expected.

Reply:
Yes, `--` would work around this but it's going to be unexpected for end users and will cost them time. This is so close; I think it's worth poking some more to see if there's a way to solve that last bit. I'll look at it tomorrow morning first thing. Thanks so much for getting it this far!

Reply:
Additional, "work around"  pypodman run centos:7 date --rm --quiet

Got this from a bug report filed... 8 years ago...

Reply:
That works today but is actually a bug; it will no longer work once (if) #1886 is merged.

Reply:
@jwhonce Lets get #1866 merged.

Reply:
Well, this has been eye-opening. Turns out, we can't do it. It's an 8-year-old [bug in `argparse`](https://bugs.python.org/issue9338). The workaround seems to be to use `store_true` for `--foo` and [add a corresponding `--no-foo` with `store_false`](https://stackoverflow.com/questions/52403065/argparse-optional-boolean).

I'm tempted to suggest going that route (and, yes, sorry, undoing this PR). Reasoning, first and foremost, is podman command-line compatibility. I've never seen anybody try `podman --rm True -i False` or any combination thereof (nor, for that matter, in docker). The pattern is always `podman --rm -i IMAGE COMMAND ...`. Second, even if someone were to try the abovementioned (using spaces), it wouldn't work: 
```
# podman run --rm false centos:7 date
Trying to pull docker.io/false:latest...Failed
...etc etc (i.e. trying to parse 'false' as an image)

# podman run --rm=false centos:7 date
(with an equals sign, it works as expected)
```
IOW, neither podman nor docker accept a space after a boolean-optional flag. They'll accept a value with an equals sign. I cannot find any way to tell `argparse` to emulate this behavior.

Given the typical usage of these flags, POLA dictates that pypodman should target the standard case. This will unfortunately impact those on the edges:
```
# pypodman run --rm=false IMAGE etc...
pypodman: error: argument --rm: ignored explicit argument 'False'
```
...but the thing is, this is a no-win situation. I favor the solution that impacts the fewest people.

As always, there may be something I've missed.

Reply:
Ya, this is always going to be a battle of go-vs-python argument parsing.  One could argue the go-style parsing is broken WRT the rest of the GNU-style world. But there's not enough white-space available to scribble arguments either way.

The only thing I'd add to Ed's suggestion is, document this for podman, pypodman, and python-podman.  Suggestion:

''While podman accepts parameters to "flag" options (e.g. ``--rm=true``), use is highly discouraged.  It is known to break compatibility with non-go based tooling, causing difficult to debug problems."

(or something to that effect).